### PR TITLE
[SPARK-41108][SPARK-41005][CONNECT][FOLLOW-UP] Deduplicate ArrowConverters codes

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR is a followup of both https://github.com/apache/spark/pull/38468 and https://github.com/apache/spark/pull/38612 that proposes to deduplicate codes in `ArrowConverters` by creating two classes `ArrowBatchIterator` and `ArrowBatchWithSchemaIterator`.  In addition, we reuse `ArrowBatchWithSchemaIterator` when creating an empty Arrow batch at `createEmptyArrowBatch`.

While I am here,
- I addressed my own comment at https://github.com/apache/spark/pull/38612#discussion_r1019838326
- Kept the support of both max rows and size. Max row size check was removed in https://github.com/apache/spark/pull/38612

### Why are the changes needed?

For better readability and maintenance.

### Does this PR introduce _any_ user-facing change?

No, both codes are not related.

### How was this patch tested?

This is refactoring so existing CI should cover.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala [hunk 1]
```diff
@@ -71,158 +71,146 @@ private[sql] class ArrowBatchStreamWriter(
...
+      rowIter: Iterator[InternalRow],
+      schema: StructType,
+      maxRecordsPerBatch: Long,
+      maxEstimatedBatchSize: Long,
+      timeZoneId: String,
+      context: TaskContext)
+    extends ArrowBatchIterator(
+      rowIter, schema, maxRecordsPerBatch, timeZoneId, context) {
+
+    private val arrowSchemaSize = SizeEstimator.estimate(arrowSchema)
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:134]: What is this supposed to achieve? This uses a lot of reflective code to figure out the size of the schema object. How is this related to the size of the batch?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala [hunk 2]
```diff
@@ -71,158 +71,146 @@ private[sql] class ArrowBatchStreamWriter(
...
+    var rowCountInLastBatch: Long = 0
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:135]: A couple of questions. Why do we need the rowcount? It is already encoded in the batch itself. If we do need the rowcount, please make the iterator return it in the next call instead relying on a side effect.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:134]: @hvanhovell, this PR is virtually pure refactoring except the couple of points I mentioned in the PR description. For the question, it came from https://github.com/apache/spark/pull/38612 to estimate the size of the batch before creating an Arrow batch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:135]: This logic is also from https://github.com/apache/spark/pull/38468, and this PR is a followup. The return type here is `Array[Byte]` that is raw binary record batch. So we cannot get the count from that unless we define other case classes to keep the row count. This class is private that is only used in the specific case.
